R/verb-pivot-longer.R

```diff
@@ -0,0 +1,517 @@
+  # Two rows with the same values in the "keys" imply that the should end up
+  # in the same row in the long format.
+
+  # 1. split spec according to "keys"
+  # 2. iterate over splitted spec and pivot_long for each part
+  # 3. row bind the results
+  spec_split <- vctrs::vec_split(spec, spec[, -(1:2)])
+  data_long_list <- purrr::map(
+    vctrs::vec_seq_along(spec_split),
+    function(idx) {
+      row <- spec_split$val[[idx]][, 1:2]
+      keys <- spec_split$key[idx, ]
+      keys$.seq <- NULL
+
+      transmute(
+        data,
+        !!!id_cols,
+        !!!keys,
+        !!!set_names(syms(row$.name), row$.value)
+      )
+    }
+  )
+
+  data_long <- purrr::reduce(data_long_list, union_all)
+
+
+  if (values_drop_na) {
+    value_cols <- unique(spec$.value)
+    # add "." variable locally so that cran doesn't complain about
+    # missing visible binding
+    . <- NULL
+
+    data_long <- dplyr::filter_at(
+      data_long,
+      value_cols,
+      dplyr::all_vars(!is.na(.))
+    )
+  }
+
+  return(data_long)
+}
+
+# The following is copy-pasted from tidyr
+# with the exception of `simplifyPieces()` which I have quickly hacked in R
+
+check_spec <- function(spec) {
+  # COPIED FROM tidyr
+
+  # Eventually should just be vec_assert() on partial_frame()
+  # Waiting for https://github.com/r-lib/vctrs/issues/198
+
+  if (!is.data.frame(spec)) {
+    stop("`spec` must be a data frame", call. = FALSE)
+  }
+
+  if (!has_name(spec, ".name") || !has_name(spec, ".value")) {
+    stop("`spec` must have `.name` and `.value` columns", call. = FALSE)
+  }
+
+  # Ensure .name and .value come first
+  vars <- union(c(".name", ".value"), names(spec))
+  spec[vars]
+}
+
+# Ensure that there's a one-to-one match from spec to data by adding
+# a special .seq variable which is automatically removed after pivotting.
+deduplicate_spec <- function(spec, df) {
+  # COPIED FROM tidyr
+
+  # Ensure each .name has a unique output identifier
+  key <- spec[setdiff(names(spec), ".name")]
+  if (vctrs::vec_duplicate_any(key)) {
+    pos <- vctrs::vec_group_loc(key)$loc
+    seq <- vector("integer", length = nrow(spec))
+    for (i in seq_along(pos)) {
+      seq[pos[[i]]] <- seq_along(pos[[i]])
+    }
+    spec$.seq <- seq
+  }
+
+  # Match spec to data, handling duplicated column names
+  col_id <- vctrs::vec_match(names(df), spec$.name)
+  has_match <- !is.na(col_id)
+
+  if (!vctrs::vec_duplicate_any(col_id[has_match])) {
+    return(spec)
+  }
+
+  spec <- vctrs::vec_slice(spec, col_id[has_match])
+  # Need to use numeric indices because names only match first
+  spec$.name <- seq_along(df)[has_match]
+
+  pieces <- vctrs::vec_split(seq_len(nrow(spec)), col_id[has_match])
+  copy <- integer(nrow(spec))
+  for (i in seq_along(pieces$val)) {
+    idx <- pieces$val[[i]]
+    copy[idx] <- seq_along(idx)
+  }
+
+  spec$.seq <- copy
+  spec
+}
+
+str_separate <- function(x, into, sep, convert = FALSE, extra = "warn", fill = "warn") {
+  # COPIED FROM tidyr
```

Comment:
Can you please stick all the copied functions into a separate file, with a comment header at top saying that they came from tidyr?